Fix incorrect duplicate conversion

Youssef1313 opened the pull request.

Youssef1313 commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -1954,7 +1954,9 @@ private void CheckMemberNameConflicts(BindingDiagnosticBag diagnostics)
                     var method = symbol as SourceMemberMethodSymbol;
-                    if (!(conversion is null))
+
+                    // We don't want to consider explicit interface implementations
+                    if (conversion is { MethodKind: MethodKind.Conversion })
```
Currently, `ConversionSignatureComparer.Comparer` appears to only be used by `conversionsAsConversions` in this code path.

Let me know if it you think it's worth adding an assert that the given `SourceUserDefinedConversionSymbol`s have `MethodKind == MethodKind.Conversion`

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:
```diff
@@ -33330,5 +33330,36 @@ public class C5 : I1<C5>
+                    static abstract implicit operator int(T t);
+                }
+
+                class Foo : ISam1<Foo>, ISam2<Foo>
```
>Foo [](http://example.com/codeflow?start=22&length=3)

Identifier like this is getting flagged by an internal tool as not politically correct. Please use more "neutral" name. "C", for example.

AlekseyTs replied:
>class Foo [](http://example.com/codeflow?start=16&length=9)

To make sure behavior is not order dependent, let's add another class with methods of interest declared in different order.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:
```diff
@@ -33330,5 +33330,36 @@ public class C5 : I1<C5>
+                    public static Foo operator +(Foo a, Foo b) => a;
+
+                    static implicit ISam1<Foo>.operator int(Foo t) => 1;
+                    static implicit ISam2<Foo>.operator int(Foo t) => 2;
```
>static implicit ISam2<Foo>.operator int(Foo t) => 2; [](http://example.com/codeflow?start=20&length=52)

I assume we don't have to have two explicit implementations in order to observe the problem. Let's simplify the scenario and get rid of the second interface

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:
```diff
@@ -33330,5 +33330,36 @@ public class C5 : I1<C5>
+            CreateCompilation("""
+                interface ISam1<T> where T : ISam1<T>
+                {
+                    static abstract T operator +(T a, T b);
```
>static abstract T operator +(T a, T b); [](http://example.com/codeflow?start=20&length=39)

For completeness, let's also add and verify a lack of conflict for:
- a unary operator
- an explicit conversion operator
- an explicit checked operator

AlekseyTs commented:
Done with review pass (commit 1)

Youssef1313 commented:
@AlekseyTs I updated the test. Since valid scenarios can't contain both implicit and explicit conversions with same signature, I introduced multiple interfaces in the test. `I1` for binary/unary + operator, `I2` for implicit conversion, and `I3` for explicit conversions.

AlekseyTs reviewed (approved):
LGTM (commit 2)

AlekseyTs commented:
@dotnet/roslyn-compiler For the second review.
